test/persistence-hooks.suite.js

```diff
@@ -2946,6 +2946,431 @@ module.exports = function(dataSource, should, connectorCapabilities) {
+            if (dataSource.connector.upsertWithWhere) {
+              ctxRecorder.records.should.eql(aCtxForModel(TestModel, {
+                where: { id: 'new-id' },
+                data: { id: 'new-id', name: 'a name' },
+              }));
+            } else {
+              ctxRecorder.records.should.eql(aCtxForModel(TestModel, {
+                instance: { id: 'new-id', name: 'a name', extra: undefined },
+                isNewInstance: true,
+              }));
+            }
+            done();
+          });
+      });
+
+      it('applies updates from `before save` hook on update', function(done) {
+        TestModel.observe('before save', function(ctx, next) {
+          ctx.data.name = 'hooked';
+          next();
+        });
+
+        TestModel.upsertWithWhere({ id: existingInstance.id },
+          { name: 'updated name' },
+          function(err, instance) {
+            if (err) return done(err);
+            instance.name.should.equal('hooked');
+            done();
+          });
+      });
+
+      it('applies updates from `before save` hook on create', function(done) {
+        TestModel.observe('before save', function(ctx, next) {
+          if (ctx.instance) {
+            ctx.instance.name = 'hooked';
+          } else {
+            ctx.data.name = 'hooked';
+          }
+          next();
+        });
+
+        TestModel.upsertWithWhere({ id: 'new-id' },
+          { id: 'new-id', name: 'new name' },
+          function(err, instance) {
+            if (err) return done(err);
+            instance.name.should.equal('hooked');
+            done();
+          });
+      });
+
+      it('validates model after `before save` hook on create', function(done) {
+        TestModel.observe('before save', invalidateTestModel());
+
+        TestModel.upsertWithWhere({ id: 'new-id' },
+          { id: 'new-id', name: 'new name' },
+          function(err, instance) {
+            (err || {}).should.be.instanceOf(ValidationError);
+            (err.details.codes || {}).should.eql({ name: ['presence'] });
+            done();
+          });
+      });
+
+      it('validates model after `before save` hook on update', function(done) {
+        TestModel.observe('before save', invalidateTestModel());
+
+        TestModel.upsertWithWhere({ id: existingInstance.id },
+          { id: existingInstance.id, name: 'updated name' },
+          function(err, instance) {
+            (err || {}).should.be.instanceOf(ValidationError);
+            (err.details.codes || {}).should.eql({ name: ['presence'] });
+            done();
+          });
+      });
+
+      it('triggers `persist` hook on create', function(done) {
+        TestModel.observe('persist', ctxRecorder.recordAndNext());
+
+        TestModel.upsertWithWhere({ id: 'new-id' },
+          { id: 'new-id', name: 'a name' },
+          function(err, instance) {
+            if (err) return done(err);
+
+            if (dataSource.connector.upsertWithWhere) {
```

Comment:
The original `upsert` method provides a slightly different set of context properties for "persist hook on create".

**connector provides upsert**

```
Model
where
data
currentInstance
hookState
options
```

**connector does not provide upsert**

```
Model
data
isNewInstance
currentInstance
hookState
options
```

(You can run `node support/describe-operation-hooks.js > hooks.html` and then view the HTML file to see comprehensive overview of operations, hooks and context properties.)

I am wondering if we can improve `upsertWithWhere` to be closer to `upsert`, most notably to provide `ctx.currentInstance`?